_posts/2018-09-06-stateless-microservices.md

```diff
@@ -0,0 +1,304 @@
+
+One of our newest users at [Wireline.io](https://www.wireline.io) raised a [feature request](https://github.com/openfaas/faas/issues/716) to enhance the HTTP route functionality of functions. Wireline wanted to write functions which could run both on [AWS Lambda](https://aws.amazon.com/free/webapps/?https://aws.amazon.com/lambda/) and on OpenFaaS without any additional changes. Around that same time the CEO of [GitLab](https://about.gitlab.com/), [Sid Sijbrandi](https://twitter.com/sytses?lang=en) reached out to learn more about Serverless and how it could be leveraged at GitLab. Sid asked me whether OpenFaaS could be used to manage both FaaS Functions and the microservices his team was more familiar with such as a Sinatra apps. He showed a particular interest in [scaling to zero](https://www.openfaas.com/blog/zero-scale/) when idle.
+
+Let's look at some background before moving into the tutorial showing how to put this into practice.
+
+## What's a Function?
+
+Before getting into what a "stateless microservice" is, let's go back to my definition of a "FaaS Function" as defined in my blog post: [Functions as a Service (FaaS) (Jan 2017)](https://blog.alexellis.io/functions-as-a-service/).
+
+Functions tend to involve:
+
+* invoking short-lived functions (Lambda has a default 1-sec timeout)
+* does not publish TCP services - often accesses third-party services or resources
+* tends to be ad-hoc/event-driven such as responding to webhooks
+* should deal gracefully with spikes in traffic
+* despite name, runs on servers backed by a fluid, non-deterministic infrastructure
+* when working well makes the complex appear simple
+* has several related concerns: infrastructure management, batching, access control, definitions, scaling & dependencies
+
+Since the original post I re-wrote my initial observations as a series of properties.
+
+Functions are:
+
+* stateless
+* emphemeral & short-lived
+* auto-scaling
+* single-purposed
+
+Functions are stateless because they do not rely on internal state-machines, stored files or sessions. Each call to the function should result in the same end-result. They don't have to be strictly idempotent, but they are easier to manage when they are.
+
+Functions are emphemeral because at any point they should be able to be replaced by any similar variant. This property means that all functions can be managed in the same way - from their health-check to their lifecycle, logging and monitoring.
+
+Functions auto-scale either from a minimum replica count to a maximum count, or even down to zero and back again to match demand or to conserve resources.
+
+Functions are single purposed, but common sense must prevail. We do not need to create a new FaaS function every time we write `function x() {}` or `def x:` in our code. A single purpose may be "CRUD for an employee record", "format this IBAN" or "identify the contents of this image using a machine learning model".
```

Comment:
To be sure I'm getting this right: 

In the case of CRUD for a single employee record, I think you're saying that the state is obviously in the SQL DB, but the function is stateless because it references that DB as static config only.

Reply:
I think the relation between the CRUD, IBAN and ML examples is that meanwhile the CRUD operation implies four tasks and other two represent single tasks, you don´t need four functions for the first, this is the allusion to "common sense".

You get it right at the stateless approach. The function doesn´t keep data in it´s container b/c if it is replaced or the functions scales, the new containers will be out of sync. But I think this is not the central idea of that paragraph.